### PR TITLE
An idea of using explicit YAML files for configurations

### PR DESCRIPTION
**Pull request by zjysteven**
## Motivation

Currently configurations are hard-coded in bash script via command-line arguments. Since YAML files offer much more structured configuration format, a better and more clear way would be to encode arguments in yaml files (as the major way for specifying configuration) and meanwhile allow users to manipulate arguments via command line (as the minor way).

## What's the intention of this PR

*I don't intend this PR to be merged*. Rather, I use it to showcase how to achieve the proposed way with minimal changes with the help of [draccus](https://github.com/dlwh/draccus) (a fork of [Pyrallis](https://github.com/eladrich/pyrallis)). I'd like to have maintainers to give a review and see if they are willing to incorporate such change into the codebase.

### The current way of specifying arguments
```bash
#!/bin/bash

if [ $# -ne 9 ]; then
    echo "Usage: $0 <DATA_PATH> <IMAGE_PATH> <LLM_VERSION> <VT_VERSION> <VT_VERSION2> <CN_VERSION> <VERSION> <TRAIN_RECIPE> <MODEL_MAX_LENGTH>"
    exit 1
fi

# Assign the arguments to variables
DATA_PATH="$1"
IMAGE_PATH="$2"
LLM_VERSION="$3"
VT_VERSION="$4"
VT_VERSION2="$5"
CN_VERSION="$6"
VERSION="$7"
TRAIN_RECIPE="$8"
MODEL_MAX_LENGTH="$9"

VT_VARIANT="${VT_VERSION#*/}"
LLM_VARIANT="${LLM_VERSION#*/}"

deepspeed --include localhost:4,5,6,7 --master_port 29501 tinyllava/train/train.py \
    --deepspeed ./scripts/zero3.json \
    --data_path  $DATA_PATH\
    --image_folder $IMAGE_PATH \
    --is_multimodal True \
    --conv_version pretrain \
    --model_name_or_path $LLM_VERSION \
    --vision_tower $VT_VERSION \
    --vision_tower2 "$VT_VERSION2" \
    --connector_type $CN_VERSION \
    --mm_vision_select_layer -2 \
    --image_aspect_ratio square \
    --attn_implementation flash_attention_2 \
    --fp16 True \
    --training_recipe $TRAIN_RECIPE \
    --tune_type_llm frozen \
    --tune_type_vision_tower frozen \
    --tune_vision_tower_from_layer 0 \
    --tune_type_connector full \
    --output_dir /mnt/data/sata/yinghu/checkpoints/llava_factory/tiny-llava-${LLM_VARIANT}-${VT_VARIANT}-${VERSION}-pretrain \
    --num_train_epochs 1 \
    --per_device_train_batch_size 32 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 2 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 24000 \
    --save_total_limit 1 \
    --learning_rate 1e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --model_max_length $MODEL_MAX_LENGTH \
    --gradient_checkpointing True \
    --dataloader_num_workers 8 \
    --lazy_preprocess True \
    --report_to tensorboard \
    --tokenizer_use_fast False \
    --run_name tiny-llava-${LLM_VARIANT}-${VT_VARIANT}-${VERSION}-pretrain
```
Currently, the [pretrain.sh](https://github.com/TinyLLaVA/TinyLLaVA_Factory/blob/main/scripts/train/pretrain.sh) hard-codes some arguments. When one wants to do pretraining with Phi, some specific arguments are coded in [train_phi.sh](https://github.com/TinyLLaVA/TinyLLaVA_Factory/blob/main/scripts/train/train_phi.sh).

### The new way of specifying arguments

My idea is instead to have a few yaml files for these arguments. In this PR, specifically there are
image
With which one simply needs to call
```bash
deepspeed --include localhost:4,5,6,7 --master_port 29501 tinyllava/train/train.py \
    --config_path ./configs/training/phi-pretrain.yml \
    --deepspeed ./configs/zero3.json
```
to initiate Phi pretraining specified by `configs/training/phi-pretrain.yml`. Inside `phi-pretrain.yml`, besides training-related arguments, it includes the model and data config by linking to `configs/model/phi.yml` and `configs/data/llava-v15-pretrain.yml`. All argument names are taken from the defined dataclasses in `tinyllava/utils/arguments.py`.

### A test comparing specified arguments of current method and proposed method

`scripts/test_args.sh` will initiate two dummy runs of loading arguments, one with the current method and the other with the proposed method, and output two files `config_original.yaml` and `config_new.yaml` that contain all specified arguments. The run with the current method would be exactly the same as `scripts/train/train_phi.sh` in terms of the arguments. [Here](https://www.diffchecker.com/VV09MNqi/) is the diff of `config_new.yaml` w.r.t. `config_original.yaml`. One can see that the only diffs are like directory names or local rank which don't matter. All other arguments match perfectly, meaning that this new way of providing configurations is valid.

### Overwrite yaml arguments via command line

This new method still allows user to specify arguments via command line input, which will overwrite the corresponding one specified in the yaml. For example, suppose one wants to change the vision tower and the learning rate, he/she can simply do so by:
```bash
deepspeed --include localhost:4,5,6,7 --master_port 29501 scripts/test_new_args.py \
    --config_path ./configs/training/phi-pretrain.yml \
    --deepspeed ./configs/zero3.json \
    --learning_rate 5e-4 \
    --model.vision_tower openai/clip-vit-large-patch14-336
```

### CODE REVIEW

**Comment by zjysteven** [comment]
Further tests show that this method will cause weird hanging of deepspeed... Closing now.
